### PR TITLE
feat: Persistent url fields for page contents

### PR DESCRIPTION
**Pull request by fsbraun**
## Summary by Sourcery

Persist page slug and overwrite URL on PageContent objects and derive PageUrl entries only from publicly visible content, updating URL handling, copying, and admin behavior accordingly.

New Features:
- Store slug and overwrite URL directly on PageContent and expose them as editable fields.
- Add Page.update_urls_from_content to generate and update PageUrl records from public content, including descendant paths, with collision protection.

Bug Fixes:
- Ensure non-public (draft) content slug changes do not affect the published page URL.
- Prevent pages without public content from resolving while keeping their slug reserved.

Enhancements:
- Refactor page and language copy operations to derive slugs and overwrite URLs from PageContent instead of PageUrl.
- Simplify ChangePageForm and PageContentAdmin by removing URL-locking logic and relying on PageContent for slug and overwrite URL.
- Optimize admin usage of slug and overwrite_url by avoiding extra PageUrl queries and using PageContent fields directly.

Tests:
- Update and extend admin and URL behavior tests to cover slug storage on PageContent, URL syncing rules for public vs non-public content, and update_urls_from_content edge cases.

## Related resources

* #8578 
* #...

## Checklist

* [ ] I have opened this pull request against ``main``
* [ ] I have added or modified the tests when changing logic
* [ ] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [ ] I have read the [contribution guidelines](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined [our Discord Server](https://discord-pr-review-channel.django-cms.org) and the channel [#pr-reviews](https://discord.com/channels/800813886689247262/1236299181761630249) to find a “pr review buddy” who is going to review my pull request.

### CODE REVIEW

**Review by vinitkumar — COMMENTED** [review]
The changes looks good to me. Let's get this merged and released. :)

**Review by vinitkumar — APPROVED** [review]

File: cms/migrations/0044_pagecontent_slug_overwrite_url.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    every PageContent object of the same page and language.
+
+    All versions of a page content implicitly shared one URL before this
+    migration, so every version receives the currently effective slug and
+    overwrite URL.
+    """
+    PageContent = apps.get_model("cms", "PageContent")
+    PageUrl = apps.get_model("cms", "PageUrl")
+
+    urls = {(url.page_id, url.language): url for url in PageUrl.objects.all().iterator()}
```
[the remark this answers is not in the document]
Comment (vinitkumar) [reply cms/migrations/0044_pagecontent_slug_overwrite_url.py]: Could this be a better implementation for this? 

```
from collections import defaultdict

  from django.db import migrations, models
  from django.db.models import Q


  BATCH_SIZE = 1000


  def _copy_urls_to_content(apps, schema_editor):
      PageContent = apps.get_model("cms", "PageContent")
      PageUrl = apps.get_model("cms", "PageUrl")
      db_alias = schema_editor.connection.alias

      last_pk = None

      while True:
          contents = PageContent.objects.using(db_alias).order_by("pk")

          if last_pk is not None:
              contents = contents.filter(pk__gt=last_pk)

          contents = list(
              contents.only("pk", "page_id", "language")[:BATCH_SIZE]
          )
          if not contents:
              break

          last_pk = contents[-1].pk

          # Grouping by language avoids fetching unrelated page/language pairs.
          page_ids_by_language = defaultdict(set)
          for content in contents:
              page_ids_by_language[content.language].add(content.page_id)

          url_filter = Q(pk__in=[])
          for language, page_ids in page_ids_by_language.items():
              url_filter |= Q(language=language, page_id__in=page_ids)

          urls = {
              (url.page_id, url.language): url
              for url in PageUrl.objects.using(db_alias)
              .filter(url_filter)
              .only("page_id", "language", "slug", "managed", "path")
          }

          updates = []
          for content in contents:
              url = urls.get((content.page_id, content.language))
              if url is None:
                  continue

              content.slug = url.slug
              content.overwrite_url = None if url.managed else url.path
              updates.append(content)

          if updates:
              PageContent.objects.using(db_alias).bulk_update(
                  updates,
                  ["slug", "overwrite_url"],
                  batch_size=BATCH_SIZE,
              )
```

This is what Codex gave me addressing the sourcery's concern.
